### PR TITLE
Change byteflag definition to determine Interface

### LINKED ISSUE #1337: Use template in a function with more than 15 entries print warnings
### Describe the bug

I have a function with a variable defined by a template with at least 2 possible types and more than 16 arguments. When I try to use Pyccel on such a function, I get warning messages about some integer variable being too large. To be exact, I get 2 messages if I have 16 arguments, 4 messages if I have 17 arguments, etc... 

These warnings appear only if the function has an argument defined by a template with at least 2 types. If a unique type is defined in the template, no warning messages appear.

### To Reproduce

```python
from pyccel.decorators import template

@template(name='T', types=['float[:,:,:,:,:]','complex[:,:,:,:,:]'])
def func(M: 'T', n1: "int64", n2: "int64", n3: "int64", n4: "int64", n5: "int64",
                 p1: "int64", p2: "int64", p3: "int64", p4: "int64", p5: "int64",
                 a1: "int64", a2: "int64", a3: "int64", a4: "int64", a5: "int64"):

    M[n1, n2, n3, n4, n5] = M[p1, p2, p3, p4, p5] + a1 + a2 + a3 + a4 + a5
    return

```
I tried to run the function in the terminal with the command line :
"pyccel file_name.py"

### Error details

```
~/Code/psydac/psydac/api/ast$ pyccel kernel_pyccel.py
/home/cato/venv/psydac/lib/python3.8/site-packages/pyccel/codegen/compiling/compilers.py:431: UserWarning: /home/cato/Code/psydac/psydac/api/ast/__pyccel__/kernel_pyccel_wrapper.c: In function ‘type_check’:
/home/cato/Code/psydac/psydac/api/ast/__pyccel__/kernel_pyccel_wrapper.c:132:1: warning: integer constant is so large that it is unsigned
  132 |         check += INT64_C(11529215046068469760);
      | ^       ~~~~~~~~~~~~~
/home/cato/Code/psydac/psydac/api/ast/__pyccel__/kernel_pyccel_wrapper.c: In function ‘add_2d_wrapper’:
/home/cato/Code/psydac/psydac/api/ast/__pyccel__/kernel_pyccel_wrapper.c:391:1: warning: integer constant is so large that it is unsigned
  391 |     else if (check == INT64_C(11682937913349382690))
      | ^   ~~~~~~~~~~~~~~~~~

  warnings.warn(UserWarning(err))

```

### Expected behavior
It should not print warning message.

### Language
The Python code is translated to Fortran.

### PR DESCRIPTION
**Pull request by EmilyBourne**
Change byte flag to depend on the possible types which can be passed to the Interface instead of depending on `flags_registry` (a list of all possible types which each argument could have). This decreases the value of the flags and fixes #1337

### CODE REVIEW

**Comment by EmilyBourne** [comment]
@tomcaruso this branch should fix your issue. Changes in this part of the code sometimes take a while to be approved as I am the only active developer comfortable with its contents and I haven't got round to writing the developer documentation yet. That said, this PR only affects a very limited (well tested) section of code. I have added documentation to the affected functions so it should be possible to review the PR without much knowledge of the context. If you want this to be released faster feel free to review it yourself to check my algorithm makes sense / my docstrings are understandable.

**Review by tomcaruso — APPROVED** [review]

File: pyccel/codegen/printing/cwrappercode.py [hunk 1]
```diff
@@ -1090,31 +1102,62 @@ def _print_Interface(self, expr):
...
+                    var_name = elem[0].name
+                    body.append(IfSection(elem[1], [AugAssign(check_var, '+' ,elem[2])]))
+                body.append(IfSection(LiteralTrue(), [PyErr_SetString('PyExc_TypeError', f'"{var_name} must be {error}"'), Return([PyccelUnarySub(LiteralInteger(1))])]))
             check_func_body += [If(*body)]
 
         check_func_body = [Assign(check_var, LiteralInteger(0))] + check_func_body
         check_func_body.append(Return([check_var]))
         # Creating check function definition
-        check_func_name = self.scope.parent_scope.get_new_name('type_check')
+        check_func_name = self.scope.parent_scope.get_new_name(f'type_check_{func_name}')
```
Comment (yguclu) [inline pyccel/codegen/printing/cwrappercode.py:1160]: Interesting... The argument `func_name` was not used in this function!

Comment (EmilyBourne) [reply pyccel/codegen/printing/cwrappercode.py:1160]: Yes! This also surprised me. But I think it makes the code clearer to name the generated type_check function something related to the function that it is handling

**Comment by EmilyBourne** [comment]
> I am having some difficulties in reverse engineering the algorithm that determines the flag and how it is used in the wrapper code. Could you please write it down explicitly, maybe in the docstring of function `_create_wrapper_check`?

Is this clearer? 707617f8

**Comment by yguclu** [comment]
> > I am having some difficulties in reverse engineering the algorithm that determines the flag and how it is used in the wrapper code. Could you please write it down explicitly, maybe in the docstring of function `_create_wrapper_check`?
> 
> Is this clearer? [707617f](https://github.com/pyccel/pyccel/commit/707617f870cd7e649676d0b58443aad606e9fba9)

Much clearer indeed! As I understand it, you are generalizing the original code, which was reserving 4 bits for each argument (I think...), by only reserving the minimum number of bits. And you use base 10 for clarity, so the bit shifts are dropped.

**Comment by EmilyBourne** [comment]
> > > I am having some difficulties in reverse engineering the algorithm that determines the flag and how it is used in the wrapper code. Could you please write it down explicitly, maybe in the docstring of function `_create_wrapper_check`?
> > 
> > 
> > Is this clearer? [707617f](https://github.com/pyccel/pyccel/commit/707617f870cd7e649676d0b58443aad606e9fba9)
> 
> Much clearer indeed! As I understand it, you are generalizing the original code, which was reserving 4 bits for each argument (I think...), by only reserving the minimum number of bits. And you use base 10 for clarity, so the bit shifts are dropped.

Yes exactly, this is slightly more general as it should still work if we use an as yet undefined type (e.g. a class). I don't use base 10 for clarity, but in order to be able to use the minimum number of bits. Previously 4 bits (16 values) were reserved to represent the 13 possible numeric types which means that the flag values 14, 15, and 16 were not utilised.

**Review by yguclu — APPROVED** [review]
In order to use this fix from Psydac we need a new release, and hence the version patch number should be updated. Would you like some other feature (i.e. PR) to be added to this version?

**Review by yguclu — APPROVED** [review]
Excellent work. Thanks a lot for the quick fix @EmilyBourne!

**Review by tomcaruso — APPROVED** [review]

**Comment by tomcaruso** [comment]
Thank you for this fix  @EmilyBourne !

**Pull request merged**